doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,562 @@
+      |  |   # relative_ptr m_start                                      |
+      |  |   # relative_ptr m_end                                        |
+      |  +---------------------------------------------------------------+
+      |
+   +--+------------------------------------------------------------+
+   | HeapAllocator                                                 |
+   |                                                               |
+   |   - expected<relative_ptr, AllocatorError> allocate(uint64_t) |
+   |   - void free(relative_ptr chunk)                             |
+   |   - template<AnotherAllocator>                                |
+   |     bool isMoveCompatible(const AnotherAllocator & rhs)       |
+   +---------------------------------------------------------------+
+
++--------------------------------+
+| template<typename Type>        |
+| IsAllocatorTypeTrait           |   Used to verify that AllocatorTypes
+|                                +---------------------------------------+
+|   static constexpr bool value; |  are compliant with Allocator Concept |
++--------------------------------+                                       |
+                                                                         |
+               +---------------------------------------------------------+-----+
+               | template<AllocatorTypes...>                                   |
+               | VariantAllocator                                              |
+               |                                                               |
+               |   - expected<relative_ptr, AllocatorError> allocate(uint64_t) |
+               |   - void free(relative_ptr chunk)                             |
+               |   - template<AnotherAllocator>                                |
+               |     bool isMoveCompatible(const AnotherAllocator & rhs)       |
+               |                                                               |
+               |   # relative_ptr m_ptrToAllocator                             |
+               +---------------------------------------------------------------+
+```
+The `Allocator` described in this diagram is similar to a C++20 concept which can
+be verified at compile time without inheritance. The verification can be realized
+with a `IsAllocatorTypeTrait` in C++14.
+The shared memory and IPC restrictions are forcing us to implement a
+VariantAllocator which provides us type independent access to the underlying
+allocator like inheritance would provide. Hereby the implementation for
+`allocate`, `free` and `isMoveCompatible` is always the same and can be
+implemented similar to the `cxx::variant` move and copy operations. This would
+make the `VariantAllocator` easier extendable since the user just has to add
+any additional allocator as type in the variadic template list. An alternative
+way of implementing this `VariantAllocator` would be with a switch statement
```

Comment:
switch is a bad alternative and essentially variant abstracts something like this.